MAINT: Speedup np.quantile.

anntzer opened the pull request:
Avoiding the unnecessary calls to moveaxis() and to any() each speed up
`np.quantile(x, .5)` (`x = np.random.rand(1000)`) by ~15% for me.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -4064,7 +4066,8 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
 
     # if any slice contained a nan, then all results on that slice are also nan
-    if np.any(n):
+    if (n.ndim == 0 and n[0]  # Fast-case for 1D.
```
I don't get this. If `n.ndim == 0` then `n[0]` raises `IndexError`.

anntzer replied:
sorry, that should have been `ndim == 1`.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -4017,8 +4017,10 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
-    # axis being sampled from `ap` to be first.
-    ap = np.moveaxis(ap, axis, 0)
+    # axis being sampled from `ap` to be first.  But we only do so if axis!=0,
+    # as moveaxis is quite slow.
```
Can we put this optimization in `moveaxis` if its really that slow?

anntzer replied:
that's slightly annoying because you'd have to write something like `if np.all(source == destination)` (as the arguments can be either scalars and vectors) but then you're paying a full array creation and iteration overhead even if source and destination were scalars to start with...
IOW the difference here is that we know the arguments are scalars.

eric-wieser replied:
I was thinking a naive approach like `if source is destination` inside moveaxis would be enough to speed this up in many cases, since the small integers are cached.

anntzer replied:
sure.

mattip commented:
> since the small integers are cached.

I am not sure if that is a language feature or an implementation detail. While PyPy did implement a work around, other implementations may not appreciate being punished for not implementing this.

eric-wieser commented:
I'd assume it's an implementation detail, and certainly wouldn't recommend having _behavior_ depend on it. Having performance rely on implementation details doesn't strike me as a particularly big problem though, especially since things like function call overhead are influenced by that anyway.

eric-wieser commented on numpy/core/numeric.py:
```diff
@@ -1451,6 +1451,10 @@ def moveaxis(a, source, destination):
     (5, 4, 3)
 
     """
+    # Fast-path for the no-op case when source and destination are small
+    # (cached) integers.  Relevant speedup e.g. for quantile.
+    if source is destination and isinstance(a, ndarray):
+        return a
```
In retrospect this probably isn't a good idea after all, as it make `np.moveaxis(a, 100, 100)` legal, and changes the value of `np.moveaxis(a, 0, 0) is a`.

anntzer replied:
So you want to move back to special casing axis=0 in quantile?  (I'm fine either way.)

eric-wieser replied:
Yeah, I guess that's probably best after all, assuming it actually makes a measurable difference

anntzer replied:
back to the old version.

charris commented:
close/reopen

charris commented:
@eric-wieser Ping.

anntzer commented:
Kindly bumping.

anntzer commented:
sure.

seberg commented:
@anntzer I merged the big refactor PR.  So it may make sense to rebase this now.  Unfortunatel (or maybe not), it may also make sense to time the whole thing again, because it is likely that there are a lot of new overheads that could be micro-optimized now.

anntzer committed:
PERF: Speedup np.quantile.

Avoiding the unnecessary calls to moveaxis() speed up
`np.quantile(x, .5)` (`x = np.random.rand(1000)`) by ~10% (although
there's a lot of variability) for me.

anntzer commented:
Rebased, only keeping the moveaxis() microoptimization (the any() one would need more refactoring, I think).  It still gives a ~10-15% speedup in the originally mentioned case, although there seems to be quite a lot of variability from run to run.

seberg commented:
Lets put it in then, thanks @anntzer.

jbrockmendel commented:
The pandas builds using npdev have started failing on some quantile tests, any chance this is involved?  https://github.com/pandas-dev/pandas/runs/4131499766?check_suite_focus=true#step:7:104

seberg commented:
No, it is due to gh-19857, though.
